File: python/0007-reverseInteger.py
class Solution:
    def reverse(self, x: int) -> int:
        ret = 0
        currNum = x if x > 0 else -x
        maxInt32 = 2 ** 31 - 1
        while currNum != 0:
            if ret > maxInt32 // 10:
                ret = 0
                break

            if ret == maxInt32 // 10:
                if x > 0:
                    if currNum % 10 > maxInt32 % 10:
                        ret = 0
                        break
                else:
                    if currNum % 10 > maxInt32 % 10 + 1:
                        ret = 0
                        break
            
            ret = ret * 10 + currNum % 10
            currNum //= 10

        if x < 0:
            ret = -ret

        return ret

s = Solution()
assert s.reverse(0) == 0
assert s.reverse(123) == 321
assert s.reverse(-123) == -321
assert s.reverse(120) == 21
assert s.reverse(99999999999) == 0
assert s.reverse(-999999999999) == 0
assert s.reverse(1534236496) == 0
assert s.reverse(-2147483412) == -2143847412

File: cpp/0007-reverseInteger.cpp
/*************************************************************************************
 * Source : https://leetcode.com/problems/reverse-integer/
 * Author : Chenglong Wei
 * Date   : 2019-04-28
 ************************************************************************************/
#include <iostream>
#include <cassert>
#include<limits.h>

class Solution {
public:
	int reverse(int x) {
        int ret = 0;
        while (x != 0) {
            int remainder = x % 10;
            x /= 10;
            
            if (ret > INT_MAX / 10 || (ret == INT_MAX / 10 && x > INT_MAX % 10)
                || ret < INT_MIN / 10 || (ret == INT_MIN / 10 && x < INT_MIN % 10)) {
                return 0;
            }

            ret = ret * 10 + remainder;
        }
        
        return ret;
    }
};

int main()
{
    Solution s;
    assert(s.reverse(1234567) == 7654321);
    assert(s.reverse(3210) == 123);
    assert(s.reverse(1534236496) == 0);
    assert(s.reverse(-2147483412) == -2143847412);
    

	return 0;
}
